File: python/844_BackspaceStringCompare.py
from unittest import TestCase
# https://leetcode.com/problems/backspace-string-compare
import itertools


class BackspaceStringCompare(TestCase):
    def backspaceCompare(self, S: 'str', T: 'str') -> 'bool':
        def gen(s):
            cnt = 0
            for ch in reversed(s):
                if ch == '#':
                    cnt += 1
                elif cnt > 0:
                    cnt -= 1
                else:
                    yield ch
        return all(a == b for a, b in itertools.zip_longest(gen(S), gen(T)))

    def test1(self):
        self.assertEqual(True, self.backspaceCompare("ab#c", "ad#c"))

    def test1(self):
        self.assertEqual(True, self.backspaceCompare("a##c", "#a#c"))

    def test1(self):
        self.assertEqual(False, self.backspaceCompare("a#c", "b"))

        

File: kotlin/src/main/kotlin/844.BackspaceStringCompare.kt
import org.junit.Assert.assertEquals
import org.junit.Test
// https://leetcode.com/problems/backspace-string-compare


class BackspaceStringCompare {
  private fun next(s: String, idx: Int): Int{
    var (i, c) = idx to 0
    while(i >= 0 && (s[i] == '#' || c > 0))
      c += if(s[i--] == '#') 1 else -1
    return i
  }
  fun backspaceCompare(S: String, T: String): Boolean {
    var (i, j) = S.length-1 to T.length-1
    while(true){
      i = next(S, i)
      j = next(T, j)
      if(i < 0 || j < 0)
        return i == j
      if(S[i--] != T[j--])
        return false
    }
  }

  @Test
  fun test1(){
    assertEquals(true, backspaceCompare("ab#c", "ad#c"))
  }

  @Test
  fun test2(){
    assertEquals(true, backspaceCompare("a##c", "#a#c"))
  }

  @Test
  fun test3(){
    assertEquals(false, backspaceCompare("a#c", "b"))
  }
}

